daisy/workflow_test.go

```diff
@@ -155,6 +155,43 @@ func TestGetSourceGCSAPIPath(t *testing.T) {
 	}
 }
 
+func TestCancelWorkflow_IsIdempotent(t *testing.T) {
+	w := testWorkflow()
+	if w.isCanceled {
+		t.Error("Didn't expect workflow to be canceled.")
+	}
+	w.CancelWorkflow()
+	w.CancelWorkflow()
+	if !w.isCanceled {
+		t.Error("Expect workflow to be canceled.")
+	}
+}
+
+func TestCancelWithReason_IsIdempotent(t *testing.T) {
+	w := testWorkflow()
+	reason := "reason"
+	w.CancelWithReason(reason)
+	w.CancelWithReason(reason)
```

Comment:
What if reasons are different? Which one should apply, the first or the second?

Reply:
Good question! There's probably an argument for keeping either the first or the last. My initial thought is first, since the last may be a timeout that occurs while the first is being cleaned up, and therefore the first would be a more specific reason.

(I updated the code and docs to reflect the first reason is kept. )